itests/migration_test.go

```diff
@@ -894,6 +900,257 @@ func TestMigrationNV23(t *testing.T) {
 	require.True(t, f090ActorPost.Code.Equals(accountNV23))
 	f090StatePost, err := clientApi.StateReadState(ctx, f090Addr, types.EmptyTSK)
 	require.NoError(t, err)
-	state := f090StatePost.State.(*account.State)
+	state := f090StatePost.State.(*account14.State)
 	require.Equal(t, state.Address, f090Addr)
 }
+
+func TestMigrationNV24(t *testing.T) {
+	req := require.New(t)
+
+	kit.QuietMiningLogs()
+
+	const (
+		nv24epoch               abi.ChainEpoch = 100
+		powerRampDurationEpochs uint64         = 200
+		blockTime                              = 10 * time.Millisecond
+	)
+	buildconstants.UpgradeTuktukPowerRampDurationEpochs = powerRampDurationEpochs
+	buildconstants.UpgradeTuktukHeight = nv24epoch
+
+	// InitialPledgeMaxPerByte is a little too low for an itest environment so gets in the way of
+	// testing the underlying calculation, so we bump it up here so it doesn't interfere.
+	miner14.InitialPledgeMaxPerByte = big.Mul(miner14.InitialPledgeMaxPerByte, big.NewInt(10)) // pre migration
+	miner15.InitialPledgeMaxPerByte = big.Mul(miner15.InitialPledgeMaxPerByte, big.NewInt(10)) // post migration
+
+	// Observe the rate of change of the pledge calculation during and after the power ramp; change
+	// is measured as a difference between the pre-FIP-0081 pledge calculation and the calculation
+	// after FIP-0081 has been applied.
+	var (
+		rateOfChangeDuringRamp []float64
+		rateOfChangeAfterRamp  []float64
+	)
+
+	testClient, _, ens := kit.EnsembleMinimal(
+		t,
+		kit.MockProofs(),
+		kit.UpgradeSchedule(
+			stmgr.Upgrade{
+				Network: network.Version23,
+				Height:  -1,
+			},
+			stmgr.Upgrade{
+				Network:   network.Version24,
+				Height:    nv24epoch,
+				Migration: filcns.UpgradeActorsV15,
+				PreMigrations: []stmgr.PreMigration{{ // should have no effect on measurements
+					PreMigration:    filcns.PreUpgradeActorsV15,
+					StartWithin:     nv24epoch / 2,
+					DontStartWithin: 2,
+					StopWithin:      2,
+				}},
+				Expensive: true,
```

Comment:
It doesn't look like this PR actually sets this struct for the main lotus node yet.  But when we do shouldn't Expensive be false ?  Looking at how this works maybe all upgrades are safe to consider as expensive?

Reply:
It's already set in https://github.com/filecoin-project/lotus/pull/12455 and I think it's always `true` where there's a state migration involved so we can do things like reject random "execute messages" API requests where they cross the migration boundary. See `git grep -i ErrExpensiveFork` for where this is used.